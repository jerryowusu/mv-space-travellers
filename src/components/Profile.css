.profile-container {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
}

.profile {
  display: grid;
  grid-auto-flow: column;
  grid-template-columns: 50% 50%;
  gap: 1rem;
  padding: 0;
  width: 80%;
  box-sizing: border-box;
}

.my-rockets,
.my-missions {
  list-style-type: none;
  padding: 0;
  border: 1px solid rgba(0, 0, 0, 0.2);
  border-radius: 10px;
}

.my-rockets li,
.my-missions li {
  display: flex;
  align-items: center;
  padding: 1rem;
  border-bottom: 1px solid rgba(0, 0, 0, 0.2);
}

.my-rockets a,
.my-missions a {
  text-decoration: none;
  color: rgb(70, 147, 240);
}

.my-rockets h3,
.my-missions h3 {
  flex-grow: 1;
}

.my-rockets li:last-child, 
.my-missions li:last-child {
  border: 0;
}
