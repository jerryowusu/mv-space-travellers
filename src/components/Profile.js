import React from 'react';
import PropTypes from 'prop-types';
import './profiles.css';

const Profile = (props) => {
  let { missions, rockets } = props;
  missions = missions.filter((mission) => mission.joined === true);
  rockets = rockets.filter((rocket) => rocket.reserved === true);
  let empty = true;
  if (missions.length > 0) empty = false;
  if (rockets.length > 0) empty = false;
  return (
    <div className="container">
      <div className="profile">
        <div className="profile-missions">
          <h2>My Missions</h2>
          <ul className="my-missions">
            {empty && (
            <li>No Missions to display.</li>
            )}
            {
                missions.map((mission) => (
                  <li key={mission.id}>
                    <h3>{mission.name}</h3>
                  </li>
                ))
              }
          </ul>
        </div>
        <div className="rockets">
          <h2>My Rockets</h2>
          <ul className="my-rockets">
            {empty && (
              <li>No rockets to display.</li>
            )}
            {
              rockets.map((rocket) => (
                <li key={rocket.id}>
                  <h3>{rocket.name}</h3>
                  <a href={rocket.wikipedia} alt="Wikipedia link" target="_blank" rel="noreferrer">Read more</a>
                </li>
              ))
            }
          </ul>
        </div>
      </div>
    </div>
  );
};

Profile.propTypes = {
  missions: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    joined: PropTypes.bool.isRequired,
  })).isRequired,
  rockets: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    image: PropTypes.string.isRequired,
    wikipedia: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    reserved: PropTypes.bool.isRequired,
  })).isRequired,
};

export default Profile;
