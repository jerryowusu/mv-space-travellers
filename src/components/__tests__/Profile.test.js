import { screen, render } from '@testing-library/react';
import '@testing-library/jest-dom';
import React from 'react';
import renderer from 'react-test-renderer';
import Profile from '../Profile';

const mockRocketsData = [
  {
    id: 'rocket-1',
    image: 'rocket-image-1',
    name: 'Rocket 1',
    description: 'rocket-description-1',
    reserved: false,
    wikipedia: 'rocket-link-1',
  },
  {
    id: 'rocket-2',
    image: 'rocket-image-2',
    name: 'Rocket 2',
    description: 'rocket-description-2',
    reserved: false,
    wikipedia: 'rocket-link-2',
  },
  {
    id: 'rocket-3',
    image: 'rocket-image-3',
    name: 'Rocket 3',
    description: 'rocket-description-3',
    reserved: false,
    wikipedia: 'rocket-link-3',
  },
];

const mockMissionsData = [
  {
    id: 'mission-1',
    image: 'mission-image-1',
    name: 'mission 1',
    description: 'mission-description-1',
    joined: false,
    wikipedia: 'mission-link-1',
  },
  {
    id: 'mission-2',
    image: 'mission-image-2',
    name: 'mission 2',
    description: 'mission-description-2',
    joined: false,
    wikipedia: 'mission-link-2',
  },
  {
    id: 'mission-3',
    image: 'mission-image-3',
    name: 'mission 3',
    description: 'mission-description-3',
    joined: false,
    wikipedia: 'mission-link-3',
  },
];

describe('Profile component tests:', () => {
  it('Component renders correctly', () => {
    const profile = renderer.create(
      <Profile
        missions={mockMissionsData}
        rockets={mockRocketsData}
      />,
    ).toJSON();
    expect(profile).toMatchSnapshot();
  });

  it('No missions message displayed correctly', () => {
    render(
      <Profile
        missions={mockMissionsData}
        rockets={mockRocketsData}
      />,
    );
    expect(screen.getByText('No Missions to display.')).toBeInTheDocument();
  });

  it('No rockets message displayed correctly', () => {
    render(
      <Profile
        missions={mockMissionsData}
        rockets={mockRocketsData}
      />,
    );
    expect(screen.getByText('No rockets to display.')).toBeInTheDocument();
  });

  it('Reserve some rockets and check if they are displayed correctly', () => {
    const reservedRockets = mockRocketsData.map((rocket) => ({ ...rocket, reserved: true }));
    const { container } = render(
      <Profile
        missions={mockMissionsData}
        rockets={reservedRockets}
      />,
    );
    expect(container.querySelector('.my-rockets').childElementCount).toBe(3);
  });

  it('Join some missions and check if they are displayed correctly', () => {
    const reservedMissions = mockMissionsData.map((mission) => ({ ...mission, joined: true }));
    const { container } = render(
      <Profile
        missions={reservedMissions}
        rockets={mockRocketsData}
      />,
    );
    expect(container.querySelector('.my-missions').childElementCount).toBe(3);
  });
});
