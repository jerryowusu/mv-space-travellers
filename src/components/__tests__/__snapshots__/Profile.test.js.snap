// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile component tests: Component renders correctly 1`] = `
<div
  className="profile-container"
>
  <div
    className="profile"
  >
    <div
      className="missions-container"
    >
      <h2>
        My Missions
      </h2>
      <ul
        className="my-missions"
      >
        <li>
          No Missions to display.
        </li>
      </ul>
    </div>
    <div
      className="rockets-container"
    >
      <h2>
        My Rockets
      </h2>
      <ul
        className="my-rockets"
      >
        <li>
          No rockets to display.
        </li>
      </ul>
    </div>
  </div>
</div>
`;
